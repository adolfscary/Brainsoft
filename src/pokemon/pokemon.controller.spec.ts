import { Test, TestingModule } from "@nestjs/testing";
import { setupDatabase, tearDownDatabase } from "../../test/database.js";
import { PrismaService } from "../prisma/prisma.service.js";
import { PokemonController } from "./pokemon.controller.js";
import { PokemonModule } from "./pokemon.module.js";

describe("PokemonController", () => {
  let controller: PokemonController;
  let prisma: PrismaService;

  beforeEach(async () => {
    await setupDatabase();
    const module: TestingModule = await Test.createTestingModule({
      imports: [PokemonModule],
    }).compile();
    controller = module.get(PokemonController);
    prisma = module.get(PrismaService);
  });

  afterEach(async () => {
    await tearDownDatabase();
  });

  it("should be defined", () => {
    expect(controller).toBeDefined();
  });

  describe("findById", () => {
    it("should get pokemon by id", async () => {
      const pokemon = await prisma.pokemon.findFirstOrThrow({
        where: { externalId: "001" },
      });
      const data = await controller.findById({ id: pokemon.id });
      expect(data?.id).toBe(pokemon.id);
      expect(data).toMatchSnapshot({
        id: expect.any(String),
        updatedAt: expect.any(Date),
        createdAt: expect.any(Date),
        evolutions: data.evolutions.map((evolution) => {
          return {
            ...evolution,
            id: expect.any(String),
            updatedAt: expect.any(Date),
            createdAt: expect.any(Date),
          };
        }),
        previousEvolutions: data.previousEvolutions.map((previousEvolution) => {
          return {
            ...previousEvolution,
            id: expect.any(String),
            updatedAt: expect.any(Date),
            createdAt: expect.any(Date),
          };
        }),
        types: data.types.map((type) => {
          return {
            ...type,
            id: expect.any(String),
            updatedAt: expect.any(Date),
            createdAt: expect.any(Date),
          };
        }),
      });
    });

    describe("findByName", () => {
      it("should get pokemon by name", async () => {
        const pokemon = await prisma.pokemon.findFirstOrThrow({
          where: { externalId: "001" },
        });
        const data = await controller.findByName({ name: pokemon.name });
        expect(data?.id).toBe(pokemon.id);
        expect(data).toMatchSnapshot({
          id: expect.any(String),
          updatedAt: expect.any(Date),
          createdAt: expect.any(Date),
          evolutions: data.evolutions.map((evolution) => {
            return {
              ...evolution,
              id: expect.any(String),
              updatedAt: expect.any(Date),
              createdAt: expect.any(Date),
            };
          }),
          previousEvolutions: data.previousEvolutions.map(
            (previousEvolution) => {
              return {
                ...previousEvolution,
                id: expect.any(String),
                updatedAt: expect.any(Date),
                createdAt: expect.any(Date),
              };
            },
          ),
          types: data.types.map((type) => {
            return {
              ...type,
              id: expect.any(String),
              updatedAt: expect.any(Date),
              createdAt: expect.any(Date),
            };
          }),
        });
      });
    });

    describe("paginate", () => {
      it("should paginate over pokemons", async () => {
        const user = await prisma.user.create({ data: { email: "test" } });
        const data = await controller.paginate({ limit: 1, page: 1 }, user);
        expect(data).toMatchSnapshot({
          data: data.data.map((one) => {
            return {
              ...one,
              id: expect.any(String),
              updatedAt: expect.any(Date),
              createdAt: expect.any(Date),
              evolutions: one.evolutions.map((evolution) => {
                return {
                  ...evolution,
                  id: expect.any(String),
                  updatedAt: expect.any(Date),
                  createdAt: expect.any(Date),
                };
              }),
              previousEvolutions: one.previousEvolutions.map(
                (previousEvolution) => {
                  return {
                    ...previousEvolution,
                    id: expect.any(String),
                    updatedAt: expect.any(Date),
                    createdAt: expect.any(Date),
                  };
                },
              ),
              types: one.types.map((type) => {
                return {
                  ...type,
                  id: expect.any(String),
                  updatedAt: expect.any(Date),
                  createdAt: expect.any(Date),
                };
              }),
            };
          }),
        });
      });
    });

    describe("paginateTypes", () => {
      it("should paginate over pokemons types", async () => {
        const data = await controller.paginateTypes({ limit: 1, page: 1 });
        expect(data).toMatchSnapshot({
          data: data.data.map((one) => {
            return {
              ...one,
              id: expect.any(String),
              updatedAt: expect.any(Date),
              createdAt: expect.any(Date),
            };
          }),
        });
      });
    });

    describe("updateFavorite", () => {
      it("should update user favorite pokemon", async () => {
        const user = await prisma.user.create({ data: { email: "test" } });
        const pokemon = await prisma.pokemon.findFirstOrThrow({
          where: { externalId: "001" },
        });
        await controller.updateFavorite(
          { id: pokemon.id },
          { favorite: true },
          user,
        );
        const updatedPokemon = await prisma.pokemon.findFirstOrThrow({
          where: { externalId: "001" },
        });
        expect(updatedPokemon.userIds).toEqual(
          expect.arrayContaining([user.id]),
        );
      });
    });
  });
});
