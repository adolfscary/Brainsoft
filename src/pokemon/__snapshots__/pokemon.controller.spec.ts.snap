// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonController findById findByName should get pokemon by name 1`] = `
{
  "attacks": {
    "fast": [
      {
        "damage": 12,
        "name": "Tackle",
        "type": "Normal",
      },
      {
        "damage": 7,
        "name": "Vine Whip",
        "type": "Grass",
      },
    ],
    "special": [
      {
        "damage": 70,
        "name": "Power Whip",
        "type": "Grass",
      },
      {
        "damage": 40,
        "name": "Seed Bomb",
        "type": "Grass",
      },
      {
        "damage": 55,
        "name": "Sludge Bomb",
        "type": "Poison",
      },
    ],
  },
  "classification": "Seed Pokémon",
  "createdAt": Any<Date>,
  "evolutionRequirements": {
    "amount": 25,
    "name": "Bulbasaur candies",
  },
  "evolutions": [
    {
      "createdAt": Any<Date>,
      "externalId": "2",
      "id": Any<String>,
      "name": "Ivysaur",
      "updatedAt": Any<Date>,
    },
    {
      "createdAt": Any<Date>,
      "externalId": "3",
      "id": Any<String>,
      "name": "Venusaur",
      "updatedAt": Any<Date>,
    },
  ],
  "externalId": "001",
  "fleeRate": 0.1,
  "height": {
    "maximum": "0.79m",
    "minimum": "0.61m",
  },
  "id": Any<String>,
  "maxCP": 951,
  "maxHP": 1071,
  "name": "Bulbasaur",
  "previousEvolutions": [],
  "resistant": [
    "Water",
    "Electric",
    "Grass",
    "Fighting",
    "Fairy",
  ],
  "types": [
    {
      "createdAt": Any<Date>,
      "id": Any<String>,
      "name": "Grass",
      "updatedAt": Any<Date>,
    },
    {
      "createdAt": Any<Date>,
      "id": Any<String>,
      "name": "Poison",
      "updatedAt": Any<Date>,
    },
  ],
  "updatedAt": Any<Date>,
  "weaknesses": [
    "Fire",
    "Ice",
    "Flying",
    "Psychic",
  ],
  "weight": {
    "maximum": "7.76kg",
    "minimum": "6.04kg",
  },
}
`;

exports[`PokemonController findById paginate should paginate over pokemons 1`] = `
{
  "data": [
    {
      "attacks": {
        "fast": [
          {
            "damage": 7,
            "name": "Pound",
            "type": "Normal",
          },
        ],
        "special": [
          {
            "damage": 65,
            "name": "Dragon Pulse",
            "type": "Dragon",
          },
          {
            "damage": 100,
            "name": "Earthquake",
            "type": "Ground",
          },
          {
            "damage": 100,
            "name": "Fire Blast",
            "type": "Fire",
          },
          {
            "damage": 80,
            "name": "Hurricane",
            "type": "Flying",
          },
          {
            "damage": 120,
            "name": "Hyper Beam",
            "type": "Normal",
          },
          {
            "damage": 85,
            "name": "Moonblast",
            "type": "Fairy",
          },
          {
            "damage": 55,
            "name": "Psychic",
            "type": "Psychic",
          },
          {
            "damage": 120,
            "name": "Solar Beam",
            "type": "Grass",
          },
          {
            "damage": 100,
            "name": "Thunder",
            "type": "Electric",
          },
        ],
      },
      "classification": "New Species Pokémon",
      "createdAt": Any<Date>,
      "evolutionRequirements": null,
      "evolutions": [],
      "externalId": "151",
      "fleeRate": 0.1,
      "height": {
        "maximum": "0.45m",
        "minimum": "0.35m",
      },
      "id": Any<String>,
      "maxCP": 3087,
      "maxHP": 3299,
      "name": "Mew",
      "previousEvolutions": [],
      "resistant": [
        "Fighting",
        "Psychic",
      ],
      "types": [
        {
          "createdAt": Any<Date>,
          "id": Any<String>,
          "name": "Psychic",
          "updatedAt": Any<Date>,
        },
      ],
      "updatedAt": Any<Date>,
      "weaknesses": [
        "Bug",
        "Ghost",
        "Dark",
      ],
      "weight": {
        "maximum": "4.5kg",
        "minimum": "3.5kg",
      },
    },
  ],
  "pageInfo": {
    "currentPage": 1,
    "isFirstPage": true,
    "isLastPage": false,
    "nextPage": 2,
    "pageCount": 151,
    "previousPage": null,
    "totalCount": 151,
  },
}
`;

exports[`PokemonController findById paginateTypes should paginate over pokemons types 1`] = `
{
  "data": [
    {
      "createdAt": Any<Date>,
      "id": Any<String>,
      "name": "Dragon",
      "updatedAt": Any<Date>,
    },
  ],
  "pageInfo": {
    "currentPage": 1,
    "isFirstPage": true,
    "isLastPage": false,
    "nextPage": 2,
    "pageCount": 17,
    "previousPage": null,
    "totalCount": 17,
  },
}
`;

exports[`PokemonController findById should get pokemon by id 1`] = `
{
  "attacks": {
    "fast": [
      {
        "damage": 12,
        "name": "Tackle",
        "type": "Normal",
      },
      {
        "damage": 7,
        "name": "Vine Whip",
        "type": "Grass",
      },
    ],
    "special": [
      {
        "damage": 70,
        "name": "Power Whip",
        "type": "Grass",
      },
      {
        "damage": 40,
        "name": "Seed Bomb",
        "type": "Grass",
      },
      {
        "damage": 55,
        "name": "Sludge Bomb",
        "type": "Poison",
      },
    ],
  },
  "classification": "Seed Pokémon",
  "createdAt": Any<Date>,
  "evolutionRequirements": {
    "amount": 25,
    "name": "Bulbasaur candies",
  },
  "evolutions": [
    {
      "createdAt": Any<Date>,
      "externalId": "2",
      "id": Any<String>,
      "name": "Ivysaur",
      "updatedAt": Any<Date>,
    },
    {
      "createdAt": Any<Date>,
      "externalId": "3",
      "id": Any<String>,
      "name": "Venusaur",
      "updatedAt": Any<Date>,
    },
  ],
  "externalId": "001",
  "fleeRate": 0.1,
  "height": {
    "maximum": "0.79m",
    "minimum": "0.61m",
  },
  "id": Any<String>,
  "maxCP": 951,
  "maxHP": 1071,
  "name": "Bulbasaur",
  "previousEvolutions": [],
  "resistant": [
    "Water",
    "Electric",
    "Grass",
    "Fighting",
    "Fairy",
  ],
  "types": [
    {
      "createdAt": Any<Date>,
      "id": Any<String>,
      "name": "Grass",
      "updatedAt": Any<Date>,
    },
    {
      "createdAt": Any<Date>,
      "id": Any<String>,
      "name": "Poison",
      "updatedAt": Any<Date>,
    },
  ],
  "updatedAt": Any<Date>,
  "weaknesses": [
    "Fire",
    "Ice",
    "Flying",
    "Psychic",
  ],
  "weight": {
    "maximum": "7.76kg",
    "minimum": "6.04kg",
  },
}
`;
